### PR TITLE
Bump the settings schema version

### LINKED ISSUE #5316: Missing settings schema version change from 0.5.0 to 0.6.0
## 🐛 Bug

In #5103 we changed the schema of napari settings, but did not bump the [schema version number](https://github.com/napari/napari/blob/1cc90a74211c6cbf3fc644da5866f9752cf1ac78/napari/settings/_napari_settings.py#L19).

If I understand the implementation correctly, I believe we also need to provide a migrator from 0.5.0 to 0.6.0, in order for the new schema version to actually be written to the new settings file.

We should probably write at least one test to cover this migration path.

This is somewhat related to #5314, as I would expect older versions of napari to respond in some useful way to future versions of the schema (e.g. by warning and ignoring them).

### PR DESCRIPTION
**Pull request by jni**
Closes #5316

In #5103 the keybinding model was moved from an internal implementation to app-model's implementation. This changed the representation of keybindings on disk, *and* it changed it in a platform-dependent way: on Mac, shortcuts that use the Command key are now written as Meta-Key, where they used to be translated to Ctrl-Key so that the preferences file could be platform-independent.

This is fixed with a migrator that swaps the control and command keys when reading keybindings from a file using the 0.5.0 schema (confusingly, the schema used in napari 0.4.19) but leaves them untouched if they come from defaults or from a 0.6.0 schema version file.

Additionally, app-model's keybinding validator does not recognise some previously-valid keys, including Control. This was *supposed* to be address by a string substitution in our shortcuts validator, but it was run *after* the keys had been validated by app-model's keybinding class, which had cleared them as invalid. This PR fixes that issue by adding `pre=True` to the pydantic validator declaration, which makes the validator run with the raw input rather than the leaf-validated input.

Finally, I added a couple of key translations to make it easier for people to hand-write keybindings: Super, Command, and Cmd are all valid modifier descriptions now.

### CODE REVIEW

**Comment by jni** [comment]
Hmmm. tbh I don't actually understand this change, ie this quote from @andy-sweet:

> If I understand the implementation correctly, I believe we also need to provide a migrator from 0.5.0 to 0.6.0, in order for the new schema version to actually be written to the new settings file.

I have no idea how to implement a "migrator" nor do I think I have time to learn how to do so before 0.5.0...

Since upgrading from 0.4.19 to 0.5.0 [works with warnings](https://github.com/napari/napari/issues/5316#issuecomment-2143577042), I'm wondering whether we can punt on this till 0.5.1...

Opinions super welcome here 😅

**Comment by psobolewskiPhD** [comment]
I don't get why the tmp setting file in the test is written with 0.5.0 schema.
Regarding the migrator, I think this schema change doesn't require one, because the new way for keybinds can handle the old way.
The problem is in the other direction?
I think we should bump the schema and maybe add something like:
```
if self.schema_version > CURRENT_SCHEMA_VERSION:
	warnings.warn("Can't downgrade settings, please use napari --reset")
```

**Comment by andy-sweet** [comment]
> Since upgrading from 0.4.19 to 0.5.0 [works with warnings](https://github.com/napari/napari/issues/5316#issuecomment-2143577042), I'm wondering whether we can punt on this till 0.5.1...

Will try to look at this today or tomorrow, but I think punting is probably fine since we've lived with this for a little while anyway.

**Comment by psobolewskiPhD** [comment]
I so rechecked with fresh env install 0.4.19, set some keybinds, and then update to 0.5.0a from pypi.
It messes up the keybindings -- everything that used Command key is broken, showing `unknown`, see:
https://github.com/napari/napari/issues/5316#issuecomment-2204506577

So we do need a migrator. The issue going from 0.4.19 to 0.5.0 is:
1. Control is now Ctrl
2. on macOS in 0.4.19 the keybinds were Control for Command and Meta for Control (unflipped basically), on 0.5.0 on macOS they are written out as Meta for Command and Ctrl for Control.
Just sed replacing Ctrl with Control makes things work on 0.5.0, but reverses command/control behavior on macOS due to point 2 above.

This won't solve the problem is in the other direction. Maybe for the future we add something like:
```
if self.schema_version > CURRENT_SCHEMA_VERSION:
	warnings.warn("Can't downgrade settings, please use napari --reset")
This won't help 0.5.0 downgrade to 0.4.19, but will help in the future in such a case.
```


Finally, I dont get why the test is failing: why the tmp setting file in the test is written with 0.5.0 schema.

**Comment by psobolewskiPhD** [comment]
Ok, what if can't figure out how to migrate keybinds properly, maybe we should add a migrator (even a no-op will update the schema version in the settings), but then have the migrator just do a warning that keybindings can't be migrated and do `napari --reset`?

**Comment by psobolewskiPhD** [comment]
I think the failed test is due to lack of migrator?
I think what is happening is:
- the settings file written at the start of the test doesn't have a schema version, so it gets 0.3.0 added as the default via:
https://github.com/napari/napari/blob/46fdf4c2adb25663fbc7cd8806c287c2c9638ef9/napari/settings/_napari_settings.py#L80-L85
- this gets migrated to 0.4.0 and 0.5.0 by the existing migrators
- without a 0.6.0 migrator, the comparison fails

**Comment by jni** [comment]
Ok the other thing I've noticed is, looking at the [implementation for the 4->5 migrator](https://github.com/napari/napari/blob/f327d1f545b88c6473f3103cc46e8beac194d1ea/napari/settings/_migrations.py#L104-L117), the *current* settings parser must be able to read the *previous* settings file. Evidently that's not the case currently.

So the path to fixing is:
- modify current reader to be able to parse old-style keybindings
- write code to convert old-style keybindings to new style ones
- use that code in the migrator

... I think I can manage, if I'm not woefully incorrect about the approach. 😅

**Comment by jni** [comment]
Posting a Q/observation in case someone has ideas... even with this PR, when I load napari and write out the settings, it *still* writes out the schema version as 0.5.0! Is there someplace else where I need to change the schema number? 🤔

**Comment by psobolewskiPhD** [comment]
@jni I think it's this: https://github.com/napari/napari/pull/7053#issuecomment-2212027739
If you add a no-op migrator:
```
@migrator('0.5.0', '0.6.0')
def v050_060(model: NapariSettings):
    """Migrate from v0.5.0 to v0.6.0

	return
```
Then it should work.

**Comment by jni** [comment]
Ok so the test that is failing (the other failure is just a side effect of this failing) is:

https://github.com/napari/napari/blob/a4fa1fa00ecee4a492daf77d7f901a4e4acc2734/napari/settings/_tests/test_settings.py#L276-L295

with the error:

```
        # make sure that the only value is now the schema version
>       assert safe_load(fake_path.read_text()) == {
            'schema_version': CURRENT_SCHEMA_VERSION
        }
E       AssertionError: assert {'schema_version': '0.6.0', 'shortcuts': {'shortcuts': {'napari:toggle_console_visibility': ['Ctrl+Shift+C'], 'napari:reset_scroll_progress': ['Meta'], 'napari:toggle_ndisplay': ['Ctrl+Y'], [...TRUNCATED]
E         
E         Common items:
E         {'schema_version': '0.6.0'}
E         Left contains 1 more item:
E         {'shortcuts': {'shortcuts': {'napari:activate_add_ellipse_mode': ['E'],
```

I git bisected and *just* bumping the schema version ([04e7e3f](https://github.com/napari/napari/pull/7053/commits/04e7e3f2e983cb6b28a69f676c04df652f68197a)) is sufficient to trigger the failure. I don't understand why, though. I'm probably doing something dumb with the version bump and/or the migrator... 🤔

**Comment by jni** [comment]
As pointed out by @psobolewskiPhD and corroborated by the fact that only the macOS tests are failing, using a no-op migration is enough to make the tests pass. But is that the correct thing here???

Will do some "real world" testing to check whether my implementation even does what I think it's doing. If I can just jettison the migrator and just do the pre-validation (so Control gets turned to Ctrl and so on before being sent to app-model), that's fine...!

**Comment by jni** [comment]
Ok, so the good news is, the migration worked like a charm! I

- install napari 0.4.19
- launched it
- set some shortcuts
- checked that they were set in the preferences file (they used the old notation — `Control-3` for 2D/3D toggle)
- upgraded napari to this branch
- launched it
- no warnings, the shortcuts I set worked, and the same settings file had been migrated to the new schema!

That's pretty magic.

The weird thing, though, is that in *both* cases, *all* the settings were written out, including the non-default settings. Obviously in the 0.4.19 case that should not have happened! Still no closer to understanding why it writes everything out under some circumstances...

Will look with fresh eyes in the morning. For now, I'll take the win that everything is working as expected! 🥳

**Comment by andy-sweet** [comment]
I spent a little time looking at this and some observations.

@jni and @psobolewskiPhD have far exceeded my knowledge of napari settings and their migrations, though that was a low bar!

The remaining failing tests on macOS also fail locally for me for similar reasons mentioned above (i.e. all the shortcuts are written). I think that's because the new migrator mutates `settings` which triggers its corresponding event which calls `NapariSettings._maybe_save`, which writes all those settings to disk (note that defaults are in memory only before this). Note that `_maybe_save` only writes settings which do not match defaults, which I assume is why the empty migrator makes all the tests pass.

I thought one promising fix was to explicitly specify the version in the input settings file, so that the migrators do not run. But they all still run, which was surprising.

Maybe another fix is to actually change the defaults for macOS, so that they don't get written? I feel like that's probably not the right solution, but I'm thinking out loud in a way that I hope is maybe helpful.

**Comment by andy-sweet** [comment]
> I thought one promising fix was to explicitly specify the version in the input settings file, so that the migrators do not run. But they all still run, which was surprising.

Actually, I did this wrong. This does make both tests pass and I think it might be the right solution. Here's the modifications from the latest commit:

```diff
diff --git a/napari/settings/_tests/test_settings.py b/napari/settings/_tests/test_settings.py
index 3cab36595..f6956e244 100644
--- a/napari/settings/_tests/test_settings.py
+++ b/napari/settings/_tests/test_settings.py
@@ -83,7 +83,8 @@ def test_settings_load_invalid_key(tmp_path, monkeypatch):
     # The invalid key will be removed
 
     fake_path = tmp_path / 'fake_path.yml'
-    data = """
+    data = f"""
+    schema_version: {CURRENT_SCHEMA_VERSION}
     application:
        non_existing_key: [1, 2]
        first_time: false
@@ -280,7 +281,7 @@ def test_settings_only_saves_non_default_values(monkeypatch, tmp_path):
     monkeypatch.setattr(os, 'environ', {})
 
     # manually get all default data and write to yaml file
-    all_data = NapariSettings(None).yaml()
+    all_data = NapariSettings(schema_version=CURRENT_SCHEMA_VERSION).yaml()
     fake_path = tmp_path / 'fake_path.yml'
     assert 'appearance' in all_data
     assert 'application' in all_data
```

I think this doesn't change the intent of the tests, but actually runs the migrator which I think we need from @jni's comments.

**Comment by jni** [comment]
Thanks @andy-sweet!!!

I'm quite happy to just make the tests pass, but it's still weird to me that napari does write out the full preferences on my machine (and evidently on all macs???). I thought maybe it was a ctrl/cmd thing but then you would expect shortcuts without modifiers not to be written?

Anyway, since it seems this issue exists on 0.4.19, I'm happy to get the tests passing, merge, and make an issue about it.

**Comment by andy-sweet** [comment]
> I'm quite happy to just make the tests pass, but it's still weird to me that napari does write out the full preferences on my machine (and evidently on all macs???). I thought maybe it was a ctrl/cmd thing but then you would expect shortcuts without modifiers not to be written?

Agreed.

> Anyway, since it seems this issue exists on 0.4.19, I'm happy to get the tests passing, merge, and make an issue about it.

Oh I didn't realize that. I will try to find some time to verify this. If that's the case, then agreed the behavior doesn't block this PR.

The only other ask here is to write a test in `test_migrations` to clarify what an example of the 0.5.0 -> 0.6.0. I can try to find some time to do that, but not sure I entirely followed all the changes.

File: napari/utils/key_bindings.py [hunk 1]
```diff
@@ -60,6 +60,8 @@ def hello_world(viewer):
 USER_KEYMAP: Mapping[str, Callable] = {}
 
 KEY_SUBS = {
+    'Command': 'Meta',
```
Comment (andy-sweet) [inline napari/utils/key_bindings.py:64]: I don't fully understand how this interacts with the 0.5.0 -> 0.6.0 migrator.

But this addition seems reasonable, assuming we want cross-platform consistency.

Comment (jni) [reply napari/utils/key_bindings.py:64]: Oh sorry this line specifically is ancillary, I'm not even sure Command is a valid shortcut currently, *but* I wanted to be able to write Command in a config file and have it work 😂 Sorry that is a red herring in review!

**Comment by andy-sweet** [comment]
For the test, I was thinking of something like the following

```python
@pytest.mark.skipif(
    bool(os.environ.get('MIN_REQ')), reason='not relevant for MIN_REQ'
)
def test_050_to_060_migration():
    settings = NapariSettings(
        schema_version='0.5.0',
        shortcuts={'shortcuts': {
            'napari:focus_axes_up': 'Option-Up',
            'napari:roll_axes': 'Control-E',
            'napari:transpose_axes': 'Command-T',
        }},
    )

    shortcuts = settings.shortcuts.shortcuts
    assert shortcuts['napari:focus_axes_up'] == 'Alt+Up'
    assert shortcuts['napari:roll_axes'] == 'Meta+E'
    assert shortcuts['napari:transpose_axes'] == 'Meta+T'
```

except not broken like this.

I.e. have one shortcut for each of the format changes we want to capture. If I understood the changes, the expected values might need to be platform-specific?

File: napari/utils/key_bindings.py (hunk 1 above)
Comment (andy-sweet) [reply napari/utils/key_bindings.py:64]: An explicit test would probably be enough for me to understand though.

**Comment by jni** [comment]
Thanks! That template definitely gets me going. Yes the resulting settings are platform specific, but that can be accounted for in the tests.

I don't think the minreq thing is required? napari is always going to be the latest version in this test... 😂

**Comment by andy-sweet** [comment]
> > Anyway, since it seems this issue exists on 0.4.19, I'm happy to get the tests passing, merge, and make an issue about it.
> 
> Oh I didn't realize that. I will try to find some time to verify this. If that's the case, then agreed the behavior doesn't block this PR.

On v0.4.19 on macOS, I also found that changing one shortcut to a non-default writes all the shortcuts.

```
napari: 0.4.19.post1
Platform: macOS-14.5-arm64-arm-64bit
System: MacOS 14.5
Python: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:51:49) [Clang 16.0.6 ]
Qt: 5.15.11
PyQt5: 5.15.10
NumPy: 1.26.4
SciPy: 1.13.1
Dask: 2024.6.0
VisPy: 0.14.3
magicgui: 0.8.3
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.7
npe2: 0.7.5

OpenGL:
  - GL version:  2.1 Metal - 88.1
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 3440x1440, scale 1.0
  - screen 2: resolution 1512x982, scale 2.0

Settings path:
  - /Users/asweet/Library/Application Support/napari/napari-0-4-19_743a5b34a671ad927c0a91364bc2c42b080e8967/settings.yaml
Plugins:
  - napari: 0.4.19.post1 (77 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-svg: 0.1.10 (2 contributions)
```

Looks like `dict(..., exclude_defaults=True, ...)` should handle this, though I'm not sure why it doesn't in v0.4.19 or v0.5.0.

**Comment by jni** [comment]
Thanks for verifying @andy-sweet! That makes me happier to release with that bug, we can make an issue and fix in upcoming releases.

**Comment by jni** [comment]
Oh fun! This test fails:

```
@pytest.mark.skipif(sys.platform != 'darwin', reason='tests migration on macs')
def test_050_to_060_migration_mac():
    """Check that Ctrl and Meta are swapped on macOS when migrating."""
    settings050 = NapariSettings(
        schema_version='0.5.0',
        shortcuts = {
            'shortcuts': {
                'napari:focus_axes_up': ['Alt-Up'],
                'napari:roll_axes': ['Control-E'],
                'napari:transpose_axes': ['Control-Meta-T'],
                'napari:paste_shape': ['V', 'Meta-T'],
            }
        },
    )
    settings060 = NapariSettings(
        schema_version='0.6.0',
        shortcuts={
            'shortcuts': {
                'napari:focus_axes_up': ['Alt-Up'],
                'napari:roll_axes': ['Meta-E'],
                'napari:transpose_axes': ['Ctrl-Meta-T'],
                'napari:paste_shape': ['V', 'Ctrl-T'],
            }
        },
    )
    assert settings050 == settings060
```

because the in memory defaults start out correct in both but they get incorrectly migrated in the 050 case. So now I need to figure out which setting is a default before migrating it...!

On the plus side I might accidentally fix the bug anyway 😂

**Comment by jni** [comment]
Ok this should be ready on tests passing!!! @andy-sweet @psobolewskiPhD if one of you could please review and approve I would be most grateful!!! 🙏😭 (I am happy of course to make any requested changes...!)

**Comment by jni** [comment]
So. Much. Green. 🌳 🌲 🌴  😍

**Comment by psobolewskiPhD** [comment]
On phone ATM but I will review before EOD.

**Review by psobolewskiPhD — APPROVED** [review]
Based on my investigations and failed attempts at addressing this issue, this Pr makes sense! I pulled it down and tested with updating in place and everything works nicely.

downgrading is still an issue, but we can't solve that here. we may want to make a change for the future to check if schema in settings.yaml is > current schema and suggest `napari --reset`

**Review by Czaki — APPROVED** [review]

**Pull request merged**